### PR TITLE
Tighten up the code that handles emitting type annotations.

### PR DESCRIPTION
**Pull request by swankjesse**
I found the previous code difficult to follow. Hopefully this is less so.

### CODE REVIEW

File: src/main/java/com/squareup/javapoet/ClassName.java [hunk 1]
```diff
@@ -32,232 +31,106 @@
...
+        concatAnnotations(annotations));
   }
 
+  @Override public ClassName withoutAnnotations() {
+    ClassName resultEnclosingClassName = enclosingClassName != null
+        ? enclosingClassName.withoutAnnotations()
+        : null;
+    return isAnnotated()
+        ? new ClassName(packageName, resultEnclosingClassName, simpleName)
+        : this;
```
Comment (JakeWharton) [inline src/main/java/com/squareup/javapoet/ClassName.java]: I'm not sure this ternary should exist. What does `TypeName.get(Map.class).annotatedWith(..).nestedClass("Entry").withoutAnnotations()` produce?

Comment (swankjesse) [reply src/main/java/com/squareup/javapoet/ClassName.java]: Just saving allocations when it's a no-op!

Comment (swankjesse) [reply src/main/java/com/squareup/javapoet/ClassName.java]: Gotta get back in a Java desktop mood. This doesn't need every allocation to be justified.

Comment (swankjesse) [reply src/main/java/com/squareup/javapoet/ClassName.java]: Ahhh, I see so it's also incorrect! Good catch!

Comment (JakeWharton) [reply src/main/java/com/squareup/javapoet/ClassName.java]: Or, put another way, this is equivalent to
```java
if (!isAnnotated()) return this;
ClassName resultEnclosingClassName = enclosingClassName != null
    ? enclosingClassName.withoutAnnotations()
    : null;
return new ClassName(packageName, resultEncosingClassName, simpleName)
```
Does `isAnnotated()` recurse on `enclosingClassName` for annotations?

Comment (swankjesse) [reply src/main/java/com/squareup/javapoet/ClassName.java]: It doesn’t. I’ll fix.

**Pull request merged**